### PR TITLE
KEYCLOAK-12177 KEYCLOAK-12178 WebAuthn: Improve usability

### PR DESCRIPTION
**Pull request by mabartos**
[KEYCLOAK-12177](https://issues.redhat.com/browse/KEYCLOAK-12177) WebAuthn authentication: improve usability by remove checkboxes

[KEYCLOAK-12178](https://issues.redhat.com/browse/KEYCLOAK-12178) WebAuthn register: Improve usability

### CODE REVIEW

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties [hunk 1]
```diff
@@ -338,7 +339,13 @@ auth-otp-form=OTP
 auth-password-form=Password
 auth-username-form=Username
 auth-username-password-form=Username and password
+
+# WebAuthn
 webauthn-authenticator=WebAuthn
 webauthn-authenticator-passwordless=WebAuthn Passwordless
+webauthn-login-title=WebAuthn Login
+webauthn-registration-title=WebAuthn security key registration
```
Comment (mposolda) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties]: Is it possible to change this to "Security Key Registration" ?

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties [hunk 2]
```diff
@@ -338,7 +339,13 @@ auth-otp-form=OTP
 auth-password-form=Password
 auth-username-form=Username
 auth-username-password-form=Username and password
+
+# WebAuthn
 webauthn-authenticator=WebAuthn
 webauthn-authenticator-passwordless=WebAuthn Passwordless
+webauthn-login-title=WebAuthn Login
```
Comment (mposolda) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties]: Is it possible to change this to "Security Key Login" ?

File: themes/src/main/resources/theme/base/login/messages/messages_en.properties [hunk 3]
```diff
@@ -338,7 +339,13 @@ auth-otp-form=OTP
...
+webauthn-registration-title=WebAuthn security key registration
+webauthn-error-title=WebAuthn Error
```
Comment (mposolda) [inline themes/src/main/resources/theme/base/login/messages/messages_en.properties]: Is it possible to change this to "Security Key Error" ?

**Review by mposolda — CHANGES_REQUESTED** [review]
@mabartos Looks nice! Thanks! I have few things:

1) Rebase will be needed. There is conflict in messages_en.properties file (which was probably introduced by me, Sorry :)

2) Is it possible to change some messages to not display "WebAuthn" to the user, but rather use terminology like "Security Key" ? We aim to do the same in some other places like "Select Authenticator" screen, new account console etc. See inline comments for more details.

3) I found a bug when testing this. Let's assume that you have authentication flow configuration like this:

Screenshot from 2020-01-31 18-09-41

and you have user with both "password" and "WebAuthn passwordless" credential. Then I am trying to login and doing this:
- Provide username
- I am on the password form. I click "try another way"
- I select "WebAuthn passwordless"
- I press "Cancel" dialog on the WebAuthn login page.
- I click "Try again" . Now I expect WebAuthn login to be retried, but instead of it, I will be redirected back to the Password form.

It is a chance that when you rebase, this bug won't happen anymore. But probably it will still happen.

**Comment by mabartos** [comment]
@mposolda Thanks for your advice!

**Comment by mposolda** [comment]
@mabartos Thanks a lot Martin! Unfortunately still few more points. Sorry...

1) Another conflict :( Now in the template.ftl file due some other commit by Jan. Do you have a chance to rebase one more and take a look at @iankko's changes and fix the conflict? Thanks!

2) Found another issue when testing - not sure if it was still there, or if it is regression since previous work. 
- Add WebAuthn Register required action to some user and login. WebAuthn Register required action is triggered
- Browser dialog about register new WebAuthn credential is shown. 
- I click "Cancel" on the browser dialog. Error page is shown with the message "Failed to register your WebAuthn Authenticator." - small issue 1 - the message is not localized and it contains word WEbAuthn on it. But this is not a blocker and can be fixed in some other commit
- I click "Try Again" . Now the page "We are sorry" is shown and the NullPointerException in server.log with stacktrace like this:
```
java.lang.NullPointerException
	at java.util.Base64$Decoder.decode(Base64.java:549)
	at org.keycloak.authentication.requiredactions.WebAuthnRegister.processAction(WebAuthnRegister.java:180)
```

I suppose it is possible to add another hidden parameter to the webauthn-error.ftl file and in the WebAuthnRegister.processAction method you can check if this parameter exists and just render the page if yes?

**Review by mposolda — APPROVED** [review]
@mabartos Thanks, I am still seeing the issue above, so will create the follow-up JIRA

**Pull request merged**